sql/src/main/java/org/apache/calcite/sql/type/CastedLiteralOperandTypeCheckers.java

```diff
@@ -0,0 +1,164 @@
+              operand,
+              iFormalOperand,
+              throwOnFailure
+          )) {
+            return false;
+          }
+
+          if (!super.checkSingleOperandType(
+              callBinding,
+              operand,
+              iFormalOperand,
+              family,
+              throwOnFailure
+          )) {
+            return false;
+          }
+
+          final SqlLiteral arg = fetchPrimitiveLiteralFromCasts(operand);
+          final BigDecimal value = arg.getValueAs(BigDecimal.class);
+          if (value.compareTo(BigDecimal.ZERO) < 0
+              || hasFractionalPart(value)) {
+            if (throwOnFailure) {
+              throw callBinding.newError(
+                  Static.RESOURCE.argumentMustBePositiveInteger(
+                      callBinding.getOperator().getName()));
+            }
+            return false;
+          }
+          if (value.compareTo(BigDecimal.valueOf(Integer.MAX_VALUE)) > 0) {
+            if (throwOnFailure) {
+              throw callBinding.newError(
+                  Static.RESOURCE.numberLiteralOutOfRange(value.toString()));
+            }
+            return false;
+          }
+          return true;
+        }
+
+        /** Returns whether a number has any fractional part.
+         *
+         * @see BigDecimal#longValueExact() */
+        private boolean hasFractionalPart(BigDecimal bd)
+        {
+          return bd.precision() - bd.scale() <= 0;
+        }
+      };
+
+  public static boolean isLiteral(SqlNode node, boolean allowCast)
+  {
+    assert node != null;
+    if (node instanceof SqlLiteral) {
+      return true;
+    }
+    if (!allowCast) {
+      return false;
+    }
+    switch (node.getKind()) {
+      case CAST:
+        // "CAST(e AS type)" is literal if "e" is literal
+        return isLiteral(((SqlCall) node).operand(0), true);
+      case MAP_VALUE_CONSTRUCTOR:
+      case ARRAY_VALUE_CONSTRUCTOR:
+        return ((SqlCall) node).getOperandList().stream()
+                               .allMatch(o -> isLiteral(o, true));
+      case DEFAULT:
+        return true; // DEFAULT is always NULL
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Fetches primitive literals from the casts, including NULL literal.
+   * It throws if the entered node isn't a primitive literal, which can be cast multiple times.
+   *
+   * Therefore, it would fail on the following types:
+   *  1. Nodes that are not of the form CAST(....(CAST LITERAL AS TYPE).....)
+   *  2. ARRAY and MAP literals. This won't be required since we are only using this method in the type checker for
+   *      primitive types
+   */
+  private static SqlLiteral fetchPrimitiveLiteralFromCasts(SqlNode node)
+  {
+    if (node == null) {
+      throw DruidException.defensive("'node' cannot be null");
+    }
+    if (node instanceof SqlLiteral) {
+      return (SqlLiteral) node;
+    }
+
+    if (SqlUtil.isLiteral(node, false)) {
+      return (SqlLiteral) node;
```

Comment:
the bot is right, the previous `if (node instance SqlLiteral)` already covers this since `SqlUtil.isLiteral` will only do the same check if `false` is set.
```
  public static boolean isLiteral(SqlNode node, boolean allowCast) {
    assert node != null;
    if (node instanceof SqlLiteral) {
      return true;
    }
    if (!allowCast) {
      return false;
    }
```

Reply:
I think the cast is incorrect here.

The  `SqlUtil.isLiteral(node, false)` would be true if the node is `ARRAY['a', 'b']` but the node won't be an instance of `SqlLiteral`. 

```
  public static boolean isLiteral(SqlNode node, boolean allowCast) {
    assert node != null;
    if (node instanceof SqlLiteral) {
      return true;
    }
    if (!allowCast) {
      return false;
    }
    switch (node.getKind()) {
    case CAST:
      // "CAST(e AS type)" is literal if "e" is literal
      return isLiteral(((SqlCall) node).operand(0), true);
    case MAP_VALUE_CONSTRUCTOR:
    case ARRAY_VALUE_CONSTRUCTOR:
      return ((SqlCall) node).getOperandList().stream()
          .allMatch(o -> isLiteral(o, true));
    case DEFAULT:
      return true; // DEFAULT is always NULL
    default:
      return false;
    }
  }

```

I'll remove this check, since it'd be false anyways.